daemon/api_snaps.go

```diff
@@ -196,21 +196,22 @@ type snapInstruction struct {
 	Action string `json:"action"`
 	Amend  bool   `json:"amend"`
 	snapRevisionOptions
-	DevMode                bool                   `json:"devmode"`
-	JailMode               bool                   `json:"jailmode"`
-	Classic                bool                   `json:"classic"`
-	IgnoreValidation       bool                   `json:"ignore-validation"`
-	IgnoreRunning          bool                   `json:"ignore-running"`
-	Unaliased              bool                   `json:"unaliased"`
-	Purge                  bool                   `json:"purge,omitempty"`
-	SystemRestartImmediate bool                   `json:"system-restart-immediate"`
-	Transaction            client.TransactionType `json:"transaction"`
-	Snaps                  []string               `json:"snaps"`
-	Users                  []string               `json:"users"`
-	ValidationSets         []string               `json:"validation-sets"`
-	QuotaGroupName         string                 `json:"quota-group"`
-	Time                   string                 `json:"time"`
-	HoldLevel              string                 `json:"hold-level"`
+	DevMode                bool                            `json:"devmode"`
+	JailMode               bool                            `json:"jailmode"`
+	Classic                bool                            `json:"classic"`
+	IgnoreValidation       bool                            `json:"ignore-validation"`
+	IgnoreRunning          bool                            `json:"ignore-running"`
+	Unaliased              bool                            `json:"unaliased"`
+	Purge                  bool                            `json:"purge,omitempty"`
+	SystemRestartImmediate bool                            `json:"system-restart-immediate"`
+	Transaction            client.TransactionType          `json:"transaction"`
+	Snaps                  []string                        `json:"snaps"`
+	Users                  []string                        `json:"users"`
+	SnapshotOptions        map[string]snap.SnapshotOptions `json:"options"`
```

Comment:
I wonder if it might be simpler to use `map[string]*snap.SnapshotOptions` here - that may make some things simpler down the line. Something like https://paste.ubuntu.com/p/pnQ3Ps5WvM/ maybe? (with that SnapshotOptions.IsEmpty could probably go away too :)

Reply:
Thanks for this question. This was one of the options I looked into. 

What I am fundamentally trying to do with as little code as possible, is to avoid non-usable - but valid - json entries, end up in snapshot metadata or REST API response. There are definitely still a case to make for `map[string]*snap.SnapshotOptions`, but I would like to share the reasoning behind not using it for your consideration:

The current approach for rest request parsing for options payload is to keep validation of json on the front-end side super light. This is perceived to be the approach in general (e.g. extra keys can be added that is used and this would not cause a problem), and is arguable pretty code efficient vs the alternative.

With this approach in mind, reality is that a valid, but non-usable json entry cannot be avoided `map[string]*snap.SnapshotOptions` alone, although it lowers number of permutations where this will happen. As example:

```
`{
      "options": {
           "snap1": {
           }
      }
}`
```
Would result in metadata showing entry:
```
...,
"options":{}
```
Which I think is not ideal
This is illustrated more completely with this example: https://goplay.tools/snippet/NPDDEGsNg62

This means that fundamentally we would still require a "empty" check, but then additionally need 1 more line of code to deal with "nil" pointers :)

Having said this, I think it is probably a better "general rule" to pointers to objects for marshalling and unmarshalling and in this case perhaps for the sake of symmetry. Otherwise it will always look questionable. So I have pretty much convinced myself to chance this after a lot of effort to explain why not to change this :)

Let me know your thoughts.

Reply:
I proposed a new approach using custom unmarshalling of the `SnapshotOptions` map.

Compared to previous solution there is a subtle difference in behaviour due to the fact that the early nil-ing options pointers for "empty" options: Previous some validation was done also on empty structures e.g. if empty options was provided for a snap what is not in the snaps list... but this will only happen for non-empty options expression.

This is inline with the current strictness of parsing, where infinite unrecognised structures can be added without affecting an otherwise, correctly described, recognised, JSON description.

Reply:
Just for demonstration purposes: https://goplay.tools/snippet/Iriw5jfDV_H
Unit tests verifies this behaviour.